### PR TITLE
IGNITE-14679 Schema serialization.

### PR DESCRIPTION
**Pull request by vladErmakov07**

### CODE REVIEW

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    protected final short version;
+
+    /** Previous version serializer. */
+    protected final AbstractSchemaSerializer previous;
+
+    /**
+     * @param ver Serializer version.
+     * @param previous Previous version serializer.
+     */
+    protected AbstractSchemaSerializer(byte ver, AbstractSchemaSerializer previous) {
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java:36]: seems should be 'short' type

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        this.version = ver;
+        this.previous = previous;
+    }
+
+    /**
+     * @param ver Serializer version.
+     */
+    protected AbstractSchemaSerializer(short ver) {
+        this.version = ver;
+        this.previous = null;
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java:46]:
```suggestion
        this(ver, null);
```

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/NativeTypes.java [hunk 3]
```diff
@@ -317,4 +318,85 @@ public static NativeType from(ColumnType type) {
...
         }
     }
+
+    /**
+     * Reads native type from byte buffer.
+     *
+     * @param buf Byte buffer.
+     * @return Native type.
+     */
+    public static NativeType fromByteBuffer(ExtendedByteBuffer buf) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/NativeTypes.java:328]: Let's move this method to SchemaSerializerImpl.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/CalcSizeByteBuffer.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.schema.marshaller.schema;
+
+/**
+ * ExtendedByteBuffer implementation that does not store any data but counts the size of the data that passed through it.
+ */
+public class CalcSizeByteBuffer implements ExtendedByteBuffer {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/CalcSizeByteBuffer.java:23]: This class looks useless.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    /**
+     * Calculates the required size from SchemaDescriptor object and allocates a new ByteBuffer with the given size.
+     *
+     * @param desc SchemaDescriptor object.
+     * @return ByteBuffer object.
+     */
+    protected ExtendedByteBuffer createByteBuffer(SchemaDescriptor desc) {
+        return ExtendedByteBufferFactory.allocate(size(desc));
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java:122]: Looks weird.
You create a fake buffer just to calculate size, and then allocate expandable buffer for serialization.
Why don't preallocate some buffer? or have a method for size calculation.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/CalcSizeByteBuffer.java (hunk 4 above)
Comment (vladErmakov07) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/CalcSizeByteBuffer.java]: Removed group of ExtendedByteBuffer classes.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java (hunk 5 above)
Comment (vladErmakov07) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/AbstractSchemaSerializer.java]: Now I have method 'size' for size calculation, thanks.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializer.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * SchemaDescriptor (De)Serializer interface.
+ */
+public interface SchemaSerializer {
+    /**
+     * Writes SchemaDescriptor object to byte buffer.
+     *
+     * @param desc SchemaDescriptor object.
+     * @return byte array.
+     */
+    byte[] bytes(SchemaDescriptor desc, ByteBuffer byteBuf);
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializer.java:33]: Missed javadoc for byteBuf param.
Why do we need return result and byteBuf at same time?

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+     *
+     * @param buf Byte buffer.
+     * @return String.
+     */
+    private String readString(ByteBuffer buf) {
+        int len = buf.getInt();
+        byte[] arr = new byte[len];
+
+        for (int i = 0; i < len; i++)
+            arr[i] = buf.get();
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:449]:
```suggestion
        buf.get(arr);
```

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializer.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+     * @param desc SchemaDescriptor object.
+     * @param byteBuf ByteBuffer object with allocated byte array.
+     */
+    void writeToBuffer(SchemaDescriptor desc, ByteBuffer byteBuf);
+
+    /**
+     * @param byteBuf Byte buffer with byte array.
+     * @return SchemaDescriptor object.
+     */
+    SchemaDescriptor value(ByteBuffer byteBuf);
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializer.java:39]:
```suggestion
    SchemaDescriptor readFrom(ByteBuffer byteBuf);
```

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializer.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * SchemaDescriptor (De)Serializer interface.
+ */
+public interface SchemaSerializer {
+    /**
+     * Writes SchemaDescriptor object to byte buffer.
+     *
+     * @param desc SchemaDescriptor object.
+     * @param byteBuf ByteBuffer object with allocated byte array.
+     */
+    void writeToBuffer(SchemaDescriptor desc, ByteBuffer byteBuf);
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializer.java:33]:
```suggestion
    void writeTo(ByteBuffer byteBuf, SchemaDescriptor desc);
```

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+        }
+    }
+
+    /**
+     * Appends precision of native type to byte buffer.
+     *
+     * @param buf Byte buffer.
+     * @param type Native type with precision.
+     */
+    private void appendPrecision(ByteBuffer buf, NativeType type) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:251]: Let's inline this method.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+        buf.putInt(precision);
+    }
+
+    /**
+     * Appends scale of native type to byte buffer.
+     *
+     * @param buf Byte buffer.
+     * @param type Native type with scale.
+     */
+    private void appendScale(ByteBuffer buf, NativeType type) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:273]: Let's inline this method.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 12]
```diff
@@ -0,0 +1,450 @@
...
+        buf.putInt(scale);
+    }
+
+    /**
+     * Appends len of native type to byte buffer.
+     *
+     * @param buf Byte buffer.
+     * @param type VarLen native type.
+     */
+    private void appendTypeLen(ByteBuffer buf, NativeType type) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:287]: Let's inline this method.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 13]
```diff
@@ -0,0 +1,450 @@
...
+    /**
+     * Appends column to byte buffer.
+     *
+     * @param buf Byte buffer.
+     * @param col Column.
+     */
+    private void appendColumn(Column col, ByteBuffer buf) {
+        buf.putInt(col.schemaIndex());
+        buf.put((byte)(col.nullable() ? 1 : 0));
+        appendString(col.name(), buf);
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:209]:
```suggestion

        appendString(col.name(), buf);        
```

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/serializer/AbstractSerializerTest.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+ */
+public class AbstractSerializerTest {
+    /**
+     * (de)Serialize schema test.
+     */
+    @Test
+    public void test() {
+        AbstractSchemaSerializer assembler = SchemaSerializerImpl.INSTANCE;
+
+        SchemaDescriptor desc = new SchemaDescriptor( 100500,
```
Comment (zstan) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/serializer/AbstractSerializerTest.java:41]: redundant whitespace

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITSchemaChangeKVViewTest.java [hunk 15]
```diff
@@ -114,6 +114,7 @@ public void testAddNewColumn() {
             kvView.put(
                 Tuple.create().set("key", 2L),
                 Tuple.create().set("valInt", 222).set("valStrNew", "str")
+
```
Comment (tledkov-gridgain) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ITSchemaChangeKVViewTest.java:117]: Empty line

**Pull request merged**